File: structural/typescript/decorator_structured/components.ts
export interface Coffee {
  getCost(): number;
  getDescription(): string;
}

export class SimpleCoffee implements Coffee {
  getCost(): number { return 2.0; }
  getDescription(): string { return 'Simple Coffee'; }
}

export interface TextComponent {
  getText(): string;
}

export class SimpleText implements TextComponent {
  constructor(private text: string) {}
  getText(): string { return this.text; }
}

export interface DataSource {
  writeData(data: string): void;
  readData(): string;
}

export class FileDataSource implements DataSource {
  private data: string = '';
  
  writeData(data: string): void {
    this.data = data;
    console.log(`Writing data to file: \${data}`);
  }
  
  readData(): string {
    console.log(`Reading data from file: \${this.data}`);
    return this.data;
  }
}

File: structural/dart/decorator_structured/components.dart
abstract class Coffee {
  double getCost();
  String getDescription();
}

class SimpleCoffee extends Coffee {
  @override
  double getCost() { return 2.0; }
  @override
  String getDescription() { return 'Simple Coffee'; }
}

abstract class TextComponent {
  String getText();
}

class SimpleText extends TextComponent {
  final String text;
  SimpleText(this.text);
  @override
  String getText() { return text; }
}

abstract class DataSource {
  void writeData(String data);
  String readData();
}

class FileDataSource extends DataSource {
  String _data = '';
  @override
  void writeData(String data) {
    _data = data;
    print('Writing data to file: \$data');
  }
  @override
  String readData() {
    print('Reading data from file: \$_data');
    return _data;
  }
}
